tests/test_interpolation.py

```diff
@@ -747,3 +749,136 @@ def fail_if_called(x: Any) -> None:
+            User(name="Bond", age=SI("${cast:str,'7'}")),
+            "age",
+            7,
+            IntegerNode,
+            id="convert_str_to_int",
+        ),
+    ],
+)
+def test_interpolation_type_validated_ok(
+    cfg: Any,
+    key: str,
+    expected_value: Any,
+    expected_node_type: Any,
+    restore_resolvers: Any,
+) -> Any:
+    def cast(t: Any, v: Any) -> Any:
+        return {"str": str, "int": int}[t](v)  # cast `v` to type `t`
+
+    def drop_last(s: str) -> str:
+        return s[0:-1]  # drop last character from string `s`
+
+    OmegaConf.register_new_resolver("cast", cast)
+    OmegaConf.register_new_resolver("drop_last", drop_last)
+
+    cfg = OmegaConf.structured(cfg)
+
+    val = cfg[key]
+    assert val == expected_value
+
+    node = cfg._get_node(key)
+    assert isinstance(node, Node)
+    assert isinstance(node._dereference_node(), expected_node_type)
+
+
+@pytest.mark.parametrize(
+    ("cfg", "key", "expected_error"),
+    [
+        pytest.param(
+            User(name="Bond", age=SI("${cast:str,seven}")),
+            "age",
+            pytest.raises(
+                ValidationError,
+                match=re.escape(
+                    "Value 'seven' could not be converted to Integer\n    full_key: age"
+                ),
+            ),
+            id="type_mismatch_resolver",
+        ),
+        pytest.param(
+            User(name="Bond", age=SI("${name}")),
+            "age",
+            pytest.raises(
+                ValidationError,
+                match=re.escape(
+                    "Value 'Bond' could not be converted to Integer\n    full_key: age"
+                ),
+            ),
+            id="type_mismatch_node_interpolation",
+        ),
+        pytest.param(
+            StructuredWithMissing(opt_num=None, num=II("opt_num")),
+            "num",
+            pytest.raises(
+                ValidationError,
+                match=re.escape("Non optional field cannot be assigned None"),
+            ),
+            id="non_optional_node_interpolation",
+        ),
+    ],
+)
+def test_interpolation_type_validated_error(
+    cfg: Any,
+    key: str,
+    expected_error: Any,
+    restore_resolvers: Any,
+) -> Any:
+    def cast(t: Any, v: Any) -> Any:
+        return {"str": str, "int": int}[t](v)  # cast `v` to type `t`
+
+    OmegaConf.register_new_resolver("cast", cast)
+
+    cfg = OmegaConf.structured(cfg)
+
+    with expected_error:
+        cfg[key]
+
+
+def test_type_validation_error_no_throw() -> None:
+    cfg = OmegaConf.structured(User(name="Bond", age=SI("${name}")))
+    bad_node = cfg._get_node("age")
+    assert bad_node._dereference_node(throw_on_resolution_failure=False) is None
+
+
+def test_resolver_output_dictconfig(restore_resolvers: Any) -> None:
+    OmegaConf.register_new_resolver("dict", lambda: {"a": 0, "b": 1})
+    cfg = OmegaConf.create({"x": "${dict:}"})
+    assert isinstance(cfg.x, DictConfig)
+    assert cfg.x.a == 0 and cfg.x.b == 1
```

Comment:
Test that interpolations in the the generated config are handled properly.
including relative interpolations (I think they should be relative to the node they are being accessed through).

```python
OmegaConf.register_new_resolver("dict", lambda: {"a": 0, "b": 1, "rel": "${.a}", "abs": "${y}"})
cfg = OmegaConf.create({"x": "${dict:}", "y": 10})
assert cfg.x == {"a": 0, "b": 1, "rel" : 0, "abc" : 10}
```

(List too).

Reply:
Done in fc1d2c3